File: core/tmp/Python/string/regular_expression_matching.py
def isMatch(s: str, p: str) -> bool:
    m, n = len(s), len(p)
    def matchs(i: int, j: int) -> bool:
        if i == 0:
            return False
        if p[j - 1] == '.':
            return True
        return s[i - 1] == p[j - 1]
    
    f = [[False] * (n + 1) for _ in range(m + 1)]
    f[0][0] = True
    for i in range(m + 1):
        for j in range(1, n + 1):
            if p[j - 1] == '*':
                f[i][j] |= f[i][j - 2]
                if matchs(i, j - 1):
                    f[i][j] |= f[i - 1][j]
            else:
                if matchs(i, j):
                    f[i][j] |= f[i - 1][j - 1]
    return f[m][n]

if __name__ == "__main__":
    print(False | True)

File: core/tmp/Java/src/com/dbc/algorithm/String/RegularExpressionMatching.java
package com.dbc.algorithm.String;

public class RegularExpressionMatching {
    public boolean isMatch(String s, String p) {
        int m = s.length(), n = p.length();
        boolean[][] f = new boolean[m + 1][n + 1];
        f[0][0] = true;
        for (int i = 0; i < m + 1; i++) {
            for (int j = 1; j < n + 1; j++) {
                if (p.charAt(j - 1) == '*') {
                    f[i][j] = f[i][j - 2];
                    if (matchs(i, j - 1, s, p)) {
                        f[i][j] = f[i][j] || f[i - 1][j];
                    }
                } else {
                    if (matchs(i, j, s, p)) {
                        f[i][j] = f[i][j]|| f[i - 1][j - 1];
                    }
                }
            }
        }
        return f[m][n];
    }

    private boolean matchs(int i, int j, String s, String p) {
        if (i == 0) {
            return false;
        }
        if (p.charAt(j - 1) == '.') {
            return true;
        }
        return s.charAt(i - 1) == p.charAt(j - 1);
    }
}
